library/core/src/option.rs

```diff
@@ -2545,3 +2546,43 @@ impl<T> Option<Option<T>> {
         }
     }
 }
+
+impl<T, const N: usize> [Option<T>; N] {
+    /// Transposes a `[Option<T>; N]` into a `Option<[T; N]>`.
+    ///
+    /// This will return `Some` only if every input element is `Some`. If any element is `None`,
+    /// all values are dropped and this returns `None`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(option_array_transpose)]
+    ///
+    /// let data = [Some(0); 1000];
+    /// let data: Option<[u8; 1000]> = data.transpose();
+    /// assert_eq!(data, Some([0; 1000]));
+    ///
+    /// let data = [Some(0), None];
+    /// let data: Option<[u8; 2]> = data.transpose();
+    /// assert_eq!(data, None);
+    /// ```
+    #[inline]
+    #[unstable(feature = "option_array_transpose", issue = "130828")]
+    pub fn transpose(self) -> Option<[T; N]> {
+        let mut array: MaybeUninit<[T; N]> = MaybeUninit::uninit();
+        let array_mut = array.as_mut_ptr();
+
+        for (i, e) in self.into_iter().enumerate() {
+            match e {
+                // SAFETY: `i` is always inbounds
+                Some(e) => unsafe {
+                    *(*array_mut).get_unchecked_mut(i) = e;
```

Comment:
You shouldn't be creating a `&mut T` here before it is initialized. You can avoid this with `array_mut.add(i).write(val)` or some of the other suggestions above.

Reply:
Interestingly Miri doesn't report any UB with `get_unchecked_mut`, [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2f25363a602f7cdae9c2cf931c0ab78a).

Reply:
This is still undecided https://github.com/rust-lang/unsafe-code-guidelines/issues/346, but [documented as UB in the standard library](https://doc.rust-lang.org/std/ptr/macro.addr_of_mut.html).